test/unit/sessions.test.js

```diff
@@ -181,11 +158,184 @@ describe('Sessions - unit/core', function () {
+        expect(session).to.have.property(serverSessionSymbol, null);
+      });
+
+      it('should start the txnNumberIncrement at zero', () => {
+        const session = new ClientSession(topology, serverSessionPool);
+        const txnNumberIncrementSymbol = getSymbolFrom(session, 'txnNumberIncrement');
+        expect(session).to.have.property(txnNumberIncrementSymbol, 0);
+      });
+    });
+
+    describe('get serverSession()', () => {
+      let serverSessionSymbol;
+      before(() => {
+        serverSessionSymbol = getSymbolFrom(
+          new ClientSession({}, serverSessionPool, {}),
+          'serverSession'
+        );
+      });
+
+      describe('from an explicit session', () => {
+        it('should always have a non-null serverSession after construction', () => {
+          const session = new ClientSession(topology, serverSessionPool, { explicit: true });
+          expect(session).to.have.a.property(serverSessionSymbol).be.an.instanceOf(ServerSession);
+          expect(session.serverSession).be.an.instanceOf(ServerSession);
+        });
+
+        it('should always have non-null serverSession even if it is ended before getter called', () => {
+          const session = new ClientSession(topology, serverSessionPool, { explicit: true });
+          session.hasEnded = true;
+          expect(session).to.have.a.property(serverSessionSymbol).be.an.instanceOf(ServerSession);
+          expect(session.serverSession).be.an.instanceOf(ServerSession);
+        });
+
+        it('should throw if the serverSession at the symbol property goes missing', () => {
+          const session = new ClientSession(topology, serverSessionPool, { explicit: true });
+          // We really want to make sure a ClientSession is not separated from its serverSession
+          session[serverSessionSymbol] = null;
+          expect(session).to.have.a.property(serverSessionSymbol).be.null;
+          expect(() => session.serverSession).throw(MongoRuntimeError);
+        });
+      });
+
+      describe('from an implicit session', () => {
+        const session = new ClientSession(topology, serverSessionPool, { explicit: false }); // make an implicit session
+
+        it('should throw if the session hasEnded before serverSession was acquired', () => {
+          expect(session).to.have.property(serverSessionSymbol, null);
+          session.hasEnded = true;
+          expect(() => session.serverSession).to.throw(MongoRuntimeError);
+        });
+
+        it('should acquire a serverSession if clientSession.hadEnded is false', () => {
+          expect(session).to.have.property(serverSessionSymbol, null);
+          session.hasEnded = false;
+          const acquireSpy = sinon.spy(serverSessionPool, 'acquire');
+          expect(session.serverSession).to.be.instanceOf(ServerSession);
+          expect(acquireSpy.calledOnce).to.be.true;
+          acquireSpy.restore();
+        });
+      });
```

Comment:
We're missing a test case where `it('should return the existing serverSession and not acquire a new one if one is already set')`

Reply:
Added!